src/backend/distributed/shared_library_init.c

```diff
@@ -1162,6 +1162,27 @@ RegisterCitusConfigVariables(void)
 		GUC_NO_SHOW_ALL,
 		NULL, NULL, NULL);
 
+	DefineCustomIntVariable(
+		"citus.values_materialization_threshold",
+		gettext_noop("Sets the maximum number of rows allowed for pushing down "
+					 "VALUES clause in multi-shard queries. If the number of "
+					 "rows exceeds the threshold, the VALUES is materialized "
+					 "via pull-push execution. When set to -1, materialization "
+					 "is disabled."),
+		gettext_noop("When the VALUES is pushed down (e.g., not materialized), "
+					 "the VALUES clause needs to be deparsed for every shard on "
+					 "the coordinator - and parsed on the workers. As this "
+					 "setting increased, the associated overhead is multiplied "
+					 "by the shard count. When materialized, the VALUES is "
+					 "deparsed and parsed once. The down side of materialization "
+					 "is that Postgres may choose a poor plan when joining "
+					 "the materialized result with tables."),
+		&ValuesMaterializationThreshold,
+		100, -1, INT_MAX,
```

Comment:
do we need -1 if 0 means "always materialize"?

Reply:
-1 is never materialize.
0 is always materialize. I added some more comments in the code

Reply:
I guess we want an "always materialize" and a "never materialize"